### ISSUE TITLE
Assist dask in choosing partition positions to avoid parser EOF error when CSV rows contain unescaped newlines

### ISSUE THREAD

**Issue by lmmx** [dask/dask#8045]
This follows discussion in #4415 on the parser error encountered (‘EOF within string’) when partitioning a file with unescaped newlines on a row. The partition naively splits on newlines (from what I understand?), which is only a valid assumption for files without unescaped within-row EOL characters.

**Would it be possible to ‘assist’ dask in identifying where to partition a CSV somehow?**

From the above issue I surmise that this is not currently a feature (but if it is and nobody thought to suggest it there please let me know).

I can imagine a function that:
- opens the file,
- seeks to the would-be partition offset minus some chunk size (expected to cover one or more lines),
- reads in a chunk
- identifies the latest row ending position within that chunk
- passes that chunk to dask as the partition point
- [repeat until file is completely partitioned]

I tried to load the [WIT](https://github.com/google-research-datasets/wit) dataset which has newlines within rows (i.e. unescaped free text, from Wikipedia image captions) and cannot get the advantage of cuDF speedup over pandas due to the problem of distributing the computation (in the way that dask achieves with partitions).

I think I’d be able to speed this up if dask could successfully partition with such a routine, but I don’t know where to put it (i.e. I don’t see the “select where to partition” aspect of dask exposed in the API, I presume it is not)

**Comment by martindurant** [comment]
As you can see, @lmmx , it's not a simple problem, and there's a reason that it's been around a long time. Your suggestions may be interesting, but hard to evaluate without some case studies. Providing custom offsets for read_block that have been obtained in some other way is certainly doable, and providing custom validation of blocks is not totally unreasonable either. However, the read_csv function is already massively overloaded with arguments, so I fear that any additions would be very carefully scrutinised.

I have an orthogonal suggestion from another project I am working on, [fsspec-reference-maker](https://github.com/intake/fsspec-reference-maker). The idea is, to make a "references" file (JSON), in which each entry corresponds to a specific bytes block of some other URL. This set of reference blocks can be viewed as a filesystem via fsspec - which dask already understands. Thus, you only need to find valid offsets and encode them, and then everything else will work out fine - without touching Dask's code. I imagine you could add a CSV module to fsspec-reference-maker which uses pandas (or `csv` directly)  to stream through the target file and identify valid offsets on the way, with any custom validation. This only need happen once, and once you have the references JSON, Dask need never know that something special is going on.

**Comment by lmmx** [comment]
I'm with you on all of the above. Regarding not overloading `read_csv`, I propose that the routine I'm outlining here can reuse the `sample_rows=10` argument.

That said, not sure about how else this check could be switched off (if it isn't switched off, then rows would be read from every offset, greedily, which may be undesirable for known-valid CSVs, e.g. those without multiline strings within rows). However I'll see what it looks like, I suspect any buffering 'cost' would be negligible so it wouldn't matter.

If no new options are to be added to `read_csv`, then partitioning a CSV with 2 or more will incur a reading cost of 10 rows of bytes per partition, using the csv module (not pandas). I hope this can be done efficiently and will need to benchmark to make sure I don't cause a performance regression.

To be able to turn it off, I would suggest adding a `sample_partition_rows=10` argument, such that `sample_partition_rows=0` would turn off this check (but for now will operate based on reusing `sample_rows`).

> **Content warning(?)** -- long documentation follows in the rest of this post, perhaps only primarily of interest during implementation. Please don't feel obliged to wade through if the level of detail is too much, I just needed to flesh it out to be sure enough how to implement it.

Please do feel free to pick up on any or none of the discussion below :smiley_cat:

## Part 1: Problem statement

I've thought about how it would be achieved, and I suggest sampling from an identical number of rows as sampled for the file head (`sample` in `read_bytes`), at the tail of each partition [backwards from each candidate newline delimiter], to consume an identical number of rows up to the row-delimiting newline that first follows the end of the partition (`sample_rows=10`).

I've also road tested some approaches to going about this, and concluded that pandas will not be able to achieve it (the Python csv module is needed instead, due to the way pandas conflates absent and empty values as "missing"/NaN).

<details><summary>Click for further details on the pandas parser (and its shortcomings)</summary>

<p>

It's important to go to the lengths of 10 rows, so as to be sure to avoid any potential edge case in which a quoted multiline field value could mimic one or more entire rows. This is already the default for sampling rows at the file head, as in the following edge case:

```py
>>> pd.read_csv(io.BytesIO(b'intA,intB,strC\n1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'))
   intA  intB                    strC
0     1     2                   hello
1     3     4                   world
2     5     6  foo\n7,8,bar\n9,10,baz
```

in which the final row [index 2] expands to:

```
5,6,"foo
7,8,bar
9,10,baz"
```

which dask could, by some odd chance, land on and, since it fulfils the requirements of the row, assume came from:

```py
>>> pd.read_csv(io.BytesIO(b'intA,intB,strC\n7,8,bar'))
   intA  intB strC
0     7     8  bar
```

which would quietly induce an error upon parsing at the next partition's start (when the `quotechar` would be encountered, giving the parser EOF error this issue opened with).

Note that this edge case relies on the validity of inconsistent quoting of a column. This corresponds to the Python csv module docs (“field” means “item in a row”):

> csv.QUOTE_MINIMAL¶
> Instructs writer objects to only quote those fields which contain special characters such as delimiter, quotechar or any of the characters in `lineterminator`.

I think sampling 10 rows from the tail of each partition per file is adequate to catch any such edge case and so to validate the next partition.

I think they should be sampled **backwards from the first row-delimiting newline after the partition end**, as is done for the head [except the head is sampled forwards, only once per file].

This check can be skipped for the final partition -- as of course there is no choice of newline delimiters following the EOF to iterate past so the EOF itself must be accepted as a row-delimiting newline, else the file is corrupt/invalid [out of scope of this feature]

- A file of 1 partition does not have to check a partition tail,
- A file of 2 partitions only has to check the first partition's tail.

It should also fail fast (i.e. move to the next `lineseparator` delimiter ASAP), by catching the same pandas parser error containing the 'EOF' substring as done when computing the `head`:

https://github.com/dask/dask/blob/e6b9e8983b26389c74036aeec615ac0e5ef34c65/dask/dataframe/io/csv.py#L593-L596

but where the file head routine raises a ValueError from the caught exception, the partition tail routine should instead move onto the next candidate `lineseparator` delimiter (with `seek_delimiter`) and retry, only failing if it reaches EOF (i.e. the routine has proceeded through the entirety of the final partition without finding a new row)

- This would mean that what was thought to be the final partition was actually entirely one of the earlier examples of a misleading column. This could only conceivably happen if the partitions were very small and the quoted column value was very long [spanning many lines]. I wouldn't expect this to be cause for concern.

Another major difference between the file head and partition tail routines: the partition tails are to be read backwards [from the first newline after the offset, and if that isn't a row-delimiting newline then the next after that, etc], whereas the file head is simply read forwards.

One aspect which I can't see any problem with (other than it's weird) is that the only sensible way to read backwards is to also reverse the bytes buffer.

Luckily I recently wrote [just such a routine](https://github.com/lmmx/wikitransp/blob/386bf26055661c9e13225c622d1844fc5b938116/src/wikitransp/scraper/buf_grep.py#L10-L101) for iterating backwards through newlines in a buffer which can be adapted for this task. I'm confident it is correct, and it is well documented and commented for clarity (I will remove the type annotations when reusing it to fit the dask house style).

- I can't think of any of the data types which would not be appropriate to do this for 
  - strings (including categoricals), integers, floats, etc are all reversible and still valid
  - datetime dtype cannot be created without converters [to my knowledge/my attempts to produce them]
  - "valid" here only means syntax, not semantics (the correct number of columns per row)

e.g. take the previous example, and pretend that the dataframe there was the sample (giving us the column names)

```py
>>> bytestr = b'intA,intB,strC\n1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'
>>> pd.read_csv(io.BytesIO(bytestr))
   intA  intB                    strC
0     1     2                   hello
1     3     4                   world
2     5     6  foo\n7,8,bar\n9,10,baz
>>> pd.read_csv(io.BytesIO(bytestr[::-1]))
  zab,01,9\nrab,8,7\noof     6     5
0                  dlrow     4     3
1                  olleh     2     1
2                   Crts  Btni  Atni
```

- Note here how the first partition in a file runs the risk of including the 1st line of a file (the header). This however can be easily detected: if the tail partition reads back to the start of the file (easy to check from offset and the tail buffer length), then only in this case would it be acceptable to read the 10 partition tail rows going forward instead of reverse (however in that case surely you'd just use the sample instead, as it'd be equivalent)

Now ignore the header row and just look at the first 2 rows of data, and pretend that the dataframe earlier was the sample (so we now have the column names)

```py
>>> rows_bytestr = b'1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'
>>> rev_rows_b = rows_bytestr[::-1]
>>> pd.read_csv(io.BytesIO(rev_rows_b), header=None).iloc[::-1, ::-1].T.reset_index(drop=True).T.reset_index(drop=True)
   0  1                       2
0  1  2                   olleh
1  3  4                   dlrow
2  5  6  zab,01,9\nrab,8,7\noof
```

Note that the column `names` should _not_ be passed to `read_csv` here: if they are, invalid lines will be salvaged (padded with NaN) whereas without providing column names parser errors are thrown as you would want in a validation step.

```py
>>> bad_rows_bytestr = rows_bytestr + b"\nJUST_ONE_COLUMN_NOT_THREE"
>>> bad_rev_b = bad_rows_bytestr[::-1]
>>> pd.read_csv(io.BytesIO(bad_rev_b), header=None).iloc[::-1, ::-1].T.reset_index(drop=True).T.reset_index(drop=True)
Traceback (most recent call last):
...
  File "pandas/_libs/parsers.pyx", line 1951, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: Expected 1 fields in line 2, saw 3
```

whereas if you provide names (let's imagine they come from the sample):

```py
>>> sample_df = pd.read_csv(io.BytesIO(bytestr))
>>> cols_rev = sample_df.columns[::-1]
>>> pd.read_csv(io.BytesIO(bad_rev_b), names=cols_rev, header=None).iloc[::-1, ::-1].T.reset_index(drop=True).T.reset_index(drop=True)
     0    1                          2
0  1.0  2.0                      olleh
1  3.0  4.0                      dlrow
2  5.0  6.0     zab,01,9\nrab,8,7\noof
3  NaN  NaN  EERHT_TON_NMULOC_ENO_TSUJ
```

- The invalid single column line gets padded with `NaN` rather than throwing an error!
- integers got quietly coerced to floats

So don't pass in column names to this validation step.

Let's say we reach our target number of rows (instead of 10, let's call it 3, so the above example from `rev_rows_b` is completed), and the next newline only has part of a row (again due to an intra-row linesep):

```py
>>> extra_line_bytestr = b'\n5,6,"foo\n'
>>> overfull_rows_bytestr = rows_bytestr + extra_line_bytestr
>>> overfull_rows_bytestr # 3 valid rows and an invalid (partial) row
b'1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"\n5,6,"foo\n'
>>> pd.read_csv(io.BytesIO(overfull_rows_bytestr), header=None)
Traceback (most recent call last):
...
  File "pandas/_libs/parsers.pyx", line 1951, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: EOF inside string starting at row 3
```

As expected, there is a parser error, and this can be overcome by stopping early:

```py
>>> pd.read_csv(io.BytesIO(overfull_rows_bytestr), header=None, nrows=3)
```

In reverse however, the invalid rows are accepted!

```
>>> rev_over_rows_b = overfull_rows_bytestr[:-1][::-1] + b"\n"
>>> rev_over_rows_b # an invalid (partial) row and 3 valid rows
b'oof",6,5\n"zab,01,9\nrab,8,7\noof",6,5\ndlrow,4,3\nolleh,2,1\n'
>>> pd.read_csv(io.BytesIO(rev_over_rows_b), header=None).iloc[::-1, ::-1].T.reset_index(drop=True).T.reset_index(drop=True)
   0  1                       2
0  1  2                   olleh
1  3  4                   dlrow
2  5  6  zab,01,9\nrab,8,7\noof
3  5  6                    oof"
```

- This appears to be a bug in pandas: in reverse, you can parse the invalid row (even though it has a lone, unquoted character)
- This could very trivially be checked for (an unescaped quotechar should not be found in any row)
  - I would rather not do this kind of validation when pandas should do it to be honest

**Since neither the start nor end delimiter can be known, I don't know if this can be so neatly resolved** (TBC)

This is not something unique to the first row, it can be pushed onto the second row

```py
>>> another_valid_row = b"a,0,0\n"
>>> pd.read_csv(io.BytesIO(another_valid_row + rev_over_rows_b), header=None)
                        0  1  2
0                       a  0  0
1                    oof"  6  5
2  zab,01,9\nrab,8,7\noof  6  5
3                   dlrow  4  3
4                   olleh  2  1
```

It does seem to be due to being directly before a row with [properly paired] `quotechar`, which perhaps confuses the parser:

```py
>>> spaced_b = rows_bytestr + b"\n" + another_valid_row[:-1] + extra_line_bytestr
>>> pd.read_csv(io.BytesIO(spaced_b), header=None)
Traceback (most recent call last):
...
  File "pandas/_libs/parsers.pyx", line 1951, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: EOF inside string starting at row 4

```

When parsed in the forward direction, the partial row is forcibly accepted: this means you do not want to decide if a row is valid or not by putting it at the start of a dataframe

I don't really understand why this is not a parser error:

```py
>>> pd.read_csv(io.BytesIO(b'foo"\n'), header=None)                 
      0
0  foo"
```

However the padding only seems to come after a valid row

```py
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\n'), header=None)
       0      1    2
0  hello  world  etc
1   foo"    NaN  NaN
>>> pd.read_csv(io.BytesIO(b'foo"\nhello,world,etc\n'), header=None)
Traceback (most recent call last):
...
  File "pandas/_libs/parsers.pyx", line 1951, in pandas._libs.parsers.raise_parser_error
pandas.errors.ParserError: Error tokenizing data. C error: Expected 1 fields in line 2, saw 3
```

The fact that they're being `NaN`-padded means they're interpreted as "blank", rather than invalid:

```py
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\n'), header=None, na_filter=False)
       0      1    2
0  hello  world  etc
1   foo"            
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)
       0      1    2
0  hello  world  etc
1   foo"            
2    etc    etc  etc
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)[2]
0    etc
1       
2    etc
Name: 2, dtype: object
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)[0][1]
'foo"'
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)[0][0]
'hello'
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)[2][0]
'etc'
>>> pd.read_csv(io.BytesIO(b'hello,world,etc\nfoo"\netc,etc,etc\n'), header=None, na_filter=False)[2][1]
''
```

- This [doesn't seem to be something pandas is capable of](https://stackoverflow.com/questions/39293118/how-to-get-pandas-to-throw-exception-or-otherwise-note-if-some-fields-are-miss#comment65955466_39293296)

---

There are 2 options then:

1. 'Manually' validate the input (check if the `quotechar` is in the dataframe values: this is always invalid!)
2. Use the Python csv module (I expect it will be more reliable at validation)

The 2nd option isn’t hard:

```py
>>> import csv
>>> r = csv.reader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n')
>>> for row in r: print(row)
['hello', 'world','etc']
['foo']
['etc', 'etc', 'etc']
>>> r = csv.reader(io.StringIO('foo"\n'))
['foo"']
>>> r = csv.reader(io.StringIO('intA,intB,strC\n1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'))
>>> for row in r: print(row)
['intA', 'intB', 'strC'],
['1', '2', 'hello'],
['3', '4', 'world']
['5', '6', 'foo\n7,8,bar\n9,10,baz']
>>> r = csv.reader(io.StringIO('1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"\n5,6,"foo\n'))
['1', '2', 'hello'],
['3', '4', 'world']
['5', '6', 'foo\n7,8,bar\n9,10,baz']
['5', '6', 'foo\n']
```

This is not actually any more valid: note that the `quotechar` that is "left open" on the last line is stripped, whereas pandas leaves it in (where it can at least be identified)

---

I expect this routine will look similar to how the `head` is made ([in the `if sample` block at the end of `read_bytes`](https://github.com/dask/dask/blob/e6b9e8983b26389c74036aeec615ac0e5ef34c65/dask/bytes/core.py#L141-L164), which is read by `reader` through a `BytesIO` buffer ⇢ `head`)

The output from this routine is as [outlined for `fsspec-reference-maker`](https://github.com/intake/fsspec-reference-maker/issues/66#issuecomment-900738600):

> So the _only_ task you need to accomplish, is to find a set of offset/size pairs that can load the data.

...and from this generate a filespec to pass to a to-be-written CSV module in `fsspec-reference-maker`.

The offsets returned from `read_byte` are a guide as to the positions _after_ which to seek to delimiters. So in that respect nothing has changed except we are refining this "seek to delimiters" part, making it "seek to `lineseparator` delimiters which are also row delimiters" (or equivalently "`lineseparator` delimiters which are not within a row column"), and validating that they are indeed row delimiters, by 10 consecutive rows [positive lookbehind condition].

Note that the routine for sampling the head, per file, works by sampling more and more of the buffer. The value of `sample` (256,000) is the minimum number of bytes to read until finding the `lineseparator`, which in `read_bytes` is referred to as the delimiter, [or if None is given], else it's the chunk size read at each iteration until finding the `lineseparator` delimiter [or the EOF]:

https://github.com/dask/dask/blob/e6b9e8983b26389c74036aeec615ac0e5ef34c65/dask/bytes/core.py#L146-L162

The same routine is needed in reverse, per partition, to identify the first valid [i.e. row-delimiting] `lineseparator`.

</p>

</details>

---

## Part 2 Overview of CSV partition validation

<details><summary>Click for an overview of the CSV partition validation task</summary>

<p>

The edge case of a shorter row is handled correctly by the CSV module:

- The `quotechar` is left in, where it can be detected

```py
>>> r = csv.reader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'))
>>> for row in r: print(row)
... 
['hello', 'world', 'etc']
['foo"']
['etc', 'etc', 'etc']
```

One possibility is a DictReader, which would presumably fail if one column (key) is not present:

- This correctly supplies `None` when it finds missing values
- This incorrectly defaults to reading the first line as a header row (when `fieldnames=None`)

```py
>>> r = csv.DictReader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'))
>>> for row in r: print(row)
... 
{'hello': 'foo"', 'world': None, 'etc': None}
{'hello': 'etc', 'world': 'etc', 'etc': 'etc'}
```

This could be avoided however by mandating the use of the `sample_df` earlier (the `head` formed from `b_sample` in dask)

```py
>>> bytestr = b'intA,intB,strC\n1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'
>>> sample_df = pd.read_csv(io.BytesIO(bytestr))
>>> r = csv.DictReader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'), fieldnames=sample_df.columns)
>>> for row in r: print(row)
... 
{'intA': 'hello', 'intB': 'world', 'strC': 'etc'}
{'intA': 'foo"', 'intB': None, 'strC': None}
{'intA': 'etc', 'intB': 'etc', 'strC': 'etc'}
```

Which is correct!

This can also handle the case when the sample does not have a header:

```py
>>> rows_bytestr = b'1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"'
>>> no_header_sample_df = pd.read_csv(io.BytesIO(rows_bytestr), header=None)
>>> r = csv.DictReader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'), fieldnames=no_header_sample_df.columns)
>>> for row in r: print(row)
... 
{0: 'hello', 1: 'world', 2: 'etc'}
{0: 'foo"', 1: None, 2: None}
{0: 'etc', 1: 'etc', 2: 'etc'}
```

The keys of the dict are integers rather than strings in that case.

The values are the fields, which can be checked for the unescaped quotechar

```py
>>> r = csv.DictReader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'), fieldnames=no_header_sample_df.columns)
>>> for row in r:
...    if any(quotechar in v for v in row.values()):
...        raise ValueError
...    print(row)
... 
{0: 'hello', 1: 'world', 2: 'etc'}
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ValueError
```

Actually it has to be a little more precise than just checking if the value contains the quotechar:

- Not preceded by the `escapechar` (so should use a regex created by an f-string
- If `doublequote` is True, then two consecutive [again, unescaped] `quotechar` should be considered as a single `quotechar`
  - This could be incorporated into the regex or could use `str.replace`

If the field has quotes around the entirety of it then these should (will) be stripped, so no need to handle that case.

So I think the regex is `rf'[^{escapechar}]{quotechar}\{1,{1 + int(doublequote)}\}'`

```py
>>> escapechar = r'\\'
>>> quotechar = '"'
>>> n_matchable = 1 + int(doublequote)
>>> re_patt_str = rf'([^{escapechar}]{quotechar})' + r'{1,' + f'{n_matchable}' + r'}'
>>> re_patt_str
'([^\\\\]"){1,2}'
>>> re_patt = re.compile(re_patt_str)
```

...but you also want:

- to match it within a str (so within `.*` either side)
- to identify the `quotechar` itself

You could therefore use a second match group inside the repetition matched group:

```py
>>> re_patt_str = rf'.*([^{escapechar}]({quotechar}))' + r'{1,' + f'{n_matchable}' + r'}.*'
>>> re_patt = re.compile(re_patt_str)
>>> re_patt.match('hello"foo').groups()
('o"', '"')
```

Note that the outer [repetition] match group includes whichever character precedes. You could also use a negative lookbehind

```py
>>> rf"(?<!({escapechar}))"
'(?<!(\\\\))'
>>> re_patt_str = rf'.*((?<!({escapechar}))({quotechar}))' + r'{1,' + f'{n_matchable}' + r'}.*'
>>> re_patt = re.compile(re_patt_str)
>>> re_patt
re.compile('.*((?<!(\\\\))(")){1,2}.*')
```

In full, it works as expected, and does not provide an unwanted non-`escapechar` in the match group (if a negative lookbehind succeeds it provides `None` as the matched entity rather than what was there instead):

```
>>> re_patt_str = rf'.*((?<!({escapechar}))({quotechar}))' + r'{1,' + f'{n_matchable}' + r'}.*'
>>> re_patt = re.compile(re_patt_str)
>>> re_patt_str
'.*((?<!(\\\\))(")){1,2}.*'
>>> re_patt.match(r'hello\"\"world').groups()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'groups'
>>> re_patt.match(r'hello\""world').groups()
('"', None, '"')
```

We can now specify this to the `csv.DictReader`, and since a match is a 3-tuple and a non-match is `None`, we can simply check against the truthiness of the match call:

```py
>>> r = csv.DictReader(io.StringIO('hello,world,etc\nfoo"\netc,etc,etc\n'), fieldnames=no_header_sample_df.columns)
>>> for row in r:
...    if any(re_patt.match(v) for v in row.values()):
...        raise ValueError
...    print(row)
{0: 'hello', 1: 'world', 2: 'etc'}
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ValueError
```

Fully validating would also mean checking that each row has the correct number of columns [i.e. consistent with the `head` sample]. In fact we can make it a self-contained demo now:

```py
import csv
import io
import re
import pandas as pd

__all__ = ["reader", "validate_dictreader", "validate_str"]

sample_header_bytestr = b'intA,intB,strC\n'
sample_df = pd.read_csv(io.BytesIO(sample_header_bytestr))

def reader(input_str, df=sample_df):
    """
    Args:
      input_str : The string to load into a new :class:`io.StringIO` buffer to create
                  the DictReader from
      df        : The DataFrame whose columns give the fieldnames fro the DictReader
    """
    buf = io.StringIO(input_str)
    return csv.DictReader(buf, fieldnames=df.columns)

def validate_dictreader(r):
    """
    Args:
      r : (:class:`csv.DictReader`) The DictReader which can be iterated to give the rows to be validated
          in the form of :class:`dict` objects (one per row).
    """
    escapechar = r"\\"
    quotechar = '"'
    doublequote = True
    n_matchable = 1 + int(doublequote)
    re_patt_str = rf'.*((?<!({escapechar}))({quotechar}))' + r'{1,' + f'{n_matchable}' + r'}.*'
    re_patt = re.compile(re_patt_str)
    validated_rows = []
    for row in r:
        if None in row.values():
            raise ValueError(f"Absent field (incomplete row) at {row=}")
        if any(re_patt.match(v) for v in row.values()):
            raise ValueError(f"{quotechar=} found in {row=}")
        validated_rows.append(row)
    for row in validated_rows:
        print(row) # Only print rows if entire DataFrame validated

def validate_str(input_str, sample_df=sample_df):
    """
    Validate the string ``input_str`` whose column names (or index if no names) are
    available through the ``sample_df`` DataFrame (sampled from the file's head).
    """
    r = reader(input_str, df=sample_df)
    validate_dictreader(r)
```

```py
>>> validate_str(input_str='hello,world,etc\nfoo"\netc,etc,etc\n')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "validate_rows.py", line 49, in validate_str
    validate_dictreader(r)
  File "validate_rows.py", line 36, in validate_dictreader
    raise ValueError(f"Absent field (incomplete row) at {row=}")
ValueError: Absent field (incomplete row) at row={'intA': 'foo"', 'intB': None, 'strC': None}
```

Note that fields being _present but empty_ is valid.

```py
>>> validate_str(input_str='1,2,etc\n3,,\n5,6,etc\n')
{'intA': '1', 'intB': '2', 'strC': 'etc'}
{'intA': '3', 'intB': '', 'strC': ''}
{'intA': '5', 'intB': '6', 'strC': 'etc'}
```

The other check I wanted of an implementation was that an 'overfull' byte string would be acceptable **if sufficient number of rows were read before the extra/partial line was read in**. Instead of 10, I suggested to pretend this sufficient number of rows was 3, and then use the following example:

```py
>>> rows_str = '1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"\n'
>>> extra_line_str = '5,6,"foo\n'
>>> overfull_rows_str = rows_str + extra_line_str
>>> overfull_rows_str # 3 valid rows and an invalid (partial) row
'1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"\n5,6,"foo\n'
>>> validate_str(overfull_rows_str)
{'intA': '1', 'intB': '2', 'strC': 'hello'}
{'intA': '3', 'intB': '4', 'strC': 'world'}
{'intA': '5', 'intB': '6', 'strC': 'foo\n7,8,bar\n9,10,baz'}
{'intA': '5', 'intB': '6', 'strC': 'foo\n'}
```

Unfortunately this does not validate correctly

</p>

</details>

---

## Part 3: How to validate correctly

<details><summary>Click to show the conclusions on partitioned CSV validation</summary>

<p>

As shown in the previous section, the csv module's `DictReader` for some reason cleans up the lone `quotechar` in the example `overfull_rows_str`

```
'1,2,hello\n3,4,world\n5,6,"foo\n7,8,bar\n9,10,baz"\n5,6,"foo\n'
```

Which should be split as:

```
1,2,hello
3,4,world
5,6,"foo\n7,8,bar\n9,10,baz"
5,6,"foo\n
```

I suspect the only way that it can be ensured to do so is to read the rows one at a time, 'greedily', not lazily (e.g. the regex `?` quantifier). The error here arises from trying to read lazily, and arriving at an invalid row as a result.

```py
>>> r = csv.reader(io.StringIO(overfull_rows_str))
>>> for row in r: print(row)
...
['1', '2', 'hello']
['3', '4', 'world']
['5', '6', 'foo\n7,8,bar\n9,10,baz']
['5', '6', 'foo\n']
```

The real take-away here is that **you can't trust the last row** (it might be unfinished), and for this reason you might want to read in 11 rows and drop the last (to get 10 rows).

**However**: note that the presence of the open `quotechar` is implicitly indicated by the presence of the newline within the row value (`'foo\n'`)
- If the row value ends in a newline, it should therefore be considered as (and validated as if) starting with a `quotechar`

This means we do not need to consider reading in an 11th line, only to check for the newline (and always making sure that the line we read in is `lineterminator`-terminated, or else we would not get that newline within the row value to check).

If we extend the previous example accordingly:

```py
>>> further_extension_str = 'one,two,three\n'
>>> further_extended_rows_str = overfull_rows_str + further_extension_str
>>> r = csv.reader(io.StringIO(further_extended_rows_str))
>>> for row in r: print(row)
... 
['1', '2', 'hello']
['3', '4', 'world']
['5', '6', 'foo\n7,8,bar\n9,10,baz']
['5', '6', 'foo\none,two,three\n']
```

Note that even though the lone `quotechar` on the 3rd row appeared to have been 'missed' (not shown in the row value), it was not missed from the row parsing, and so when further rows are given, the parser is not 'tricked' into believing the values to be on another row, but [correctly] conjoins them with an intra-row newline (`'foo\n'` ⇒ `'foo\none,two,three\n'`).

```py
>>> overfull_recovery_str = 'etc"'
>>> recovered_overfull_rows_str = further_extended_rows_str + overfull_extension_str
>>> r = csv.reader(io.StringIO(recovered_overfull_rows_str))
>>> for row in r: print(row)
...
['1', '2', 'hello']
['3', '4', 'world']
['5', '6', 'foo\n7,8,bar\n9,10,baz']
['5', '6', 'foo\none,two,three\none,two,three\netc']
```

Note that here, the final value has no newline (`lineseparator`) at the end of it, indicating that the multiline string value is closed.

- Also note that naturally, this will still apply when reading in reverse (since from either direction a single `quotechar` will be picked up). The same problems would be encountered with multiline field values, just in reverse.

The last issue to address is the possibility of entering within a multiline string, e.g. consider if the parser picked up _after_ the first of these lines (becoming 'tricked' into treating the lines in all-caps as rows when in fact they are all a continuation of the line `'1,2,"hello\n'`):

```
1,2,"hello
A,B,C
D,E,F
G,H,I
J,K,L"
3,4,etc
```

Of course when all of these lines are present we can validate it:

```py
>>> full_str = '1,2,"hello\nA,B,C\nD,E,F\nG,H,I\nJ,K,L"\n3,4,etc\n'
>>> validate_str(full_str)
{'intA': '1', 'intB': '2', 'strC': 'hello\nA,B,C\nD,E,F\nG,H,I\nJ,K,L'}
{'intA': '3', 'intB': '4', 'strC': 'etc'}
```

but if we omit the first line:

```py
>>> trick_str = full_str[11:]
>>> trick_str
'A,B,C\nD,E,F\nG,H,I\nJ,K,L"\n3,4,etc\n'
>>> validate_str(trick_str)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "validate_rows.py", line 49, in validate_str
    validate_dictreader(r)
  File "validate_rows.py", line 38, in validate_dictreader
    raise ValueError(f"{quotechar=} found in {row=}")
ValueError: quotechar='"' found in row={'intA': 'J', 'intB': 'K', 'strC': 'L"'}
```

Then we need to read through 4 lines until meeting the lone [closing] `quotechar` which raises the parser error.
This is as expected, and something to bear in mind when interpreting the meaning of this error. It also gives further reason _not_ to reduce the number of rows to sample from the partition tails, as too few risk not exitting the end of the multiline quoted string the parser started within.

To illustrate, here's the case when only 3 lines are iterated through (not reaching the `quotechar` on the 4th):

```py
>>> r = reader(trick_str)
>>> escapechar = r"\\"
>>> quotechar = '"'
>>> doublequote = True
>>> n_matchable = 1 + int(doublequote)
>>> re_patt_str = rf'.*((?<!({escapechar}))({quotechar}))' + r'{1,' + f'{n_matchable}' + r'}.*'
>>> re_patt = re.compile(re_patt_str)
>>> validated_rows = []
>>> for row_idx, row in enumerate(r):
>>>     if row_idx == 3: break
>>>     if None in row.values():
>>>         raise ValueError(f"Absent field (incomplete row) at {row=}")
>>>     if any(re_patt.match(v) for v in row.values()):
>>>         raise ValueError(f"{quotechar=} found in {row=}")
>>>     validated_rows.append(row)
>>>
>>> for row in validated_rows:
>>>     print(row) # Only print rows if entire DataFrame validated
{'intA': 'A', 'intB': 'B', 'strC': 'C'}
{'intA': 'D', 'intB': 'E', 'strC': 'F'}
{'intA': 'G', 'intB': 'H', 'strC': 'I'}
{'intA': '3', 'intB': '4', 'strC': 'etc'}
```

i.e. it validates the lines incorrectly into rows, as it couldn't distinguish. The only way to avoid this is a sufficiently large (and an assumption that this will be a rare occurrence in real world data).

---

The csv module therefore provides solutions to all of the problems relevant to this task:

1. Ability to indicate when more bytes are needed to complete a row (which can be read in from a buffer iteratively)
  - Indicated by a newline at the end of the field value [always at the end of a row]
2. Ability to indicate when the start position was within a row
  - Indicated by a `quotechar` within a field value (anywhere)
3. Ability to indicate when the row is not a valid row
  - Indicated by any `None` values in the `DictReader` field values

The function given earlier (`validate_str`) is sufficient to do this, except it doesn't check for newlines (as in a real implementation, instead of throwing a ValueError you'd want to read in more bytes, which could be done without discarding the rows you already have(?)

> If the row value ends in a newline, it should therefore be considered as (and validated as if) starting with a quotechar

</p>

</details>

**Comment by martindurant** [comment]
This is quite a lot of information to take in, so before attempting that, a quick question, which of these does your procedure do:
- given a random offset in a CSV file, find the location of valid newline character that is truly a row terminator
- scan a CSV from the beginning to find the locations of all true row-terminating newlines

**Comment by lmmx** [comment]
Apologies for any information overload! I just wanted it in one place after having reached a conclusion I was confident in.

I misread your question, it does the former. It does not read an entire file!

It would only need to read an entire file if the file was a single row extending over multiple lines (if it was a single row on one line or otherwise in one partition then it would be assumed valid and nothing computed: `[0, EOF)` is assumed a valid partition range).

It does not need to proceed from the start (the algorithm is not sequential/recursive). It can take a random offset in a CSV file (the only other required info is the number of columns).

- Above I assumed the number of columns would require the `head` sample DataFrame, but in fact we could reduce that requirement to just the number of columns.

In practice however it will run once on an entire file due to the [unlikely] possibility of a partition to extend through one or more entire subsequent partitions (thus 'collapsing'/consuming them). To have confidence in the uniqueness of the row-terminator offsets you should compute them all (not just one)

- Potentially in parallel but I don't know if I can assume multiprocessing is permitted so would just do it sequentially.

I plan to reuse the _node visitor/translator callback_ pattern from the `fsspec_reference_maker.hdf` module to keep consistency across the library.

(P.S. Forgot to mention: routine described above will be done by the `fsspec_reference_maker.csv` module so that dask "need never know", as you recommended. But I left the notes here as the purpose of that routine is dask-specific.)

**Comment by lmmx** [comment]
I'm not sure if the dask `read_bytes` interface can be changed (if a new argument is added to it then it may break existing code that uses it), so I propose popping a kwarg at the very start of `read_bytes`,

```py
    sample_tail_rows = kwargs.pop("sample_tail_rows", 0)
```

passed only by `read_pandas` https://github.com/dask/dask/blob/e6b9e8983b26389c74036aeec615ac0e5ef34c65/dask/dataframe/io/csv.py#L547

and then the remaining kwargs would be valid for the `storage_options`, as used now.

This variable is needed to change https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/bytes/core.py#L115-L116

into a condition block

```py
if sample_tail_rows > 0:
    # calculate offsets the CSV specific way)
else:
    off = list(range(0, size, blocksize)) 
    length = [blocksize] * len(off) 
```

Would this be acceptable?
